sympy/polys/ring_series.py

```diff
@@ -1748,3 +1787,118 @@ def rs_compose_add(p1, p2):
     if dp:
         q = q*x**dp
     return q
+
+
+_convert_func = {
+        'sin': 'rs_sin',
+        'cos': 'rs_cos',
+        'exp': 'rs_exp',
+        'tan': 'rs_tan'
+        }
+
+def rs_min_pow(expr, series_rs, a):
+    series = 0
+    n = 2
+    while series == 0:
+        series = _rs_series(expr, series_rs, a, n)
+        n *= 2
+    R = series.ring
+    a = R(a)
+    i = R.gens.index(a)
+    return min(series, key=lambda t: t[i])[i]
+
+
+def _rs_series(expr, series_rs, a, prec):
+    # XXX Use _parallel_dict_from_expr instead of sring.
+
+    args = expr.args
+    R = series_rs.ring
+    if not any(arg.has(Function) for arg in args) and not expr.is_Function:
+        return series_rs
+    if not expr.has(a):
+        return series_rs
+    elif expr.is_Function:
+        arg = args[0]
+        if len(args) > 1:
+            raise NotImplementedError
+        R1, series = sring(arg, domain=QQ, expand=False)
+        series_inner = _rs_series(arg, series, a, prec)
+        R = (R.compose(R1)).compose(series_inner.ring)
+        series_inner = series_inner.set_ring(R)
+        series = eval(_convert_func[str(expr.func)])(series_inner,
+            R(a), prec)
+        return series
+    elif expr.is_Mul:
+        n = len(args)
+
+        for arg in args:    # XXX Looks redundant
+            R1, _ = sring(arg, expand=False)
+            R = R.compose(R1)
+        min_pows = list(map(rs_min_pow, args, [R(arg) for arg in args],
+            [a]*len(args)))
+        sum_pows = sum(min_pows)
+        series = R(1)
+        for i in range(n):
+            _series = _rs_series(args[i], R(args[i]), a, prec - sum_pows +
+                min_pows[i])
+            R = R.compose(_series.ring)
+            _series = _series.set_ring(R)
+            series = series.set_ring(R)
+            series *= _series
+        series = rs_trunc(series, R(a), prec)
+        return series
+    elif expr.is_Add:
+        n = len(args)
+        series = R(0)
+        for i in range(n):
+            _series = _rs_series(args[i], R(args[i]), a, prec)
+            R = R.compose(_series.ring)
+            _series = _series.set_ring(R)
+            series = series.set_ring(R)
+            series += _series
+        return series
+    elif expr.is_Pow:
+        R1, _ = sring(expr.base, domain=QQ, expand=False)
+        R = R.compose(R1)
+        series_inner = _rs_series(expr.base, R(expr.base), a, prec)
+        return rs_pow(series_inner, expr.exp, series_inner.ring(a), prec)
+    elif isinstance(expr, Expr) and expr.is_constant():
+        return sring(expr, domain=QQ, expand=False)[1]
+    else:
+        raise NotImplementedError
+
+def rs_series(expr, a, prec):
```

Comment:
Since this is a higher level function, it needs a docstring with documentation and a few doctests showing how to use it.